fendermint/app/src/app.rs

```diff
@@ -620,11 +623,42 @@ where
         *guard = Some(state);
 
         let response = match result {
-            Err(e) => invalid_check_tx(AppError::InvalidEncoding, e.description),
+            Err(e) => {
+                emit(MpoolReceivedInvalidMessage {
+                    reason: "InvalidEncoding",
+                    description: e.description.as_ref(),
+                });
+                invalid_check_tx(AppError::InvalidEncoding, e.description)
+            }
             Ok(result) => match result {
-                Err(IllegalMessage) => invalid_check_tx(AppError::IllegalMessage, "".to_owned()),
-                Ok(Err(InvalidSignature(d))) => invalid_check_tx(AppError::InvalidSignature, d),
-                Ok(Ok(ret)) => to_check_tx(ret),
+                Err(IllegalMessage) => {
+                    emit(MpoolReceivedInvalidMessage {
+                        reason: "IllegalMessage",
+                        description: "",
+                    });
```

Comment:
1. What do you think about merging the `MpoolReceivedInvalidMessage` and `MpoolReceived` events into a single type (`MpoolReceived`), where `accept` determines whether it was accepted or not?
2. I would merge the reason and the description into a single field: `reason`, and [`format!()`](https://doc.rust-lang.org/std/macro.format.html) the original reason and description into it. You can then leverage the derived `Debug` implementation of the `AppError` enum to drop the inline strings (e.g. `"InvalidEncoding"`).
3. You can initialize a mutable struct at the top, update it throughout the nested `match`es, and emit it at the end, so the code is less pedantic.

Reply:
I wanted to follow your suggestion, but that would require making a lot of fields in MpoolReceived optional. This is particularly necessary for cases like InvalidEncoding, where the transaction fails to parse the raw bytes. However, I suppose we can live with the optional types.

Reply:
Alright, I have implemented it using options.